### PR TITLE
Create validation and release workflows for nuget packages

### PR DESCRIPTION
**Pull request by fiedukow**
This adds a build and nuget package creation workflows to our CI.

The `bcny-build-nuget-package` is a callable workflow that is used for both PRs and releases to build `libsodium` and package form it. It's then used in validate workflow that is used for PRs and in release workflow that adds an extra job of uploading the artifacts to nuget at the end.

I tested that by temporarily adding step of uploading to PR checks - it seems to upload artifacts as expected.

### CODE REVIEW

File: .github/workflows/bcny-build-nuget-package.yml
Comment (compnerd) [inline .github/workflows/bcny-build-nuget-package.yml:9]: Any reason to not do both simultaneously as a matrix?

File: .github/workflows/bcny-build-nuget-package.yml
Comment (compnerd) [inline .github/workflows/bcny-build-nuget-package.yml:43]: Is this a dll or static version?  I also wonder if this actually works with non-link linkers.

File: .github/workflows/bcny-build-nuget-package.yml
Comment (compnerd) [inline .github/workflows/bcny-build-nuget-package.yml]: The `.exp` file is not used, we should drop that.

File: .github/workflows/bcny-build-nuget-package.yml
Comment (compnerd) [inline .github/workflows/bcny-build-nuget-package.yml]: Lets hold off on the shared version until we deem it necessary.

File: .github/workflows/bcny-build-nuget-package.yml
Comment (compnerd) [inline .github/workflows/bcny-build-nuget-package.yml]: Lets do `lib` to maintain uniformity with all the other nuget packages.

File: .github/workflows/bcny-build-nuget-package.yml
Comment (compnerd) [inline .github/workflows/bcny-build-nuget-package.yml]: Similar, lets just package the version that we intend to use currently.

**Review by compnerd — COMMENTED** [review]
That is an interesting approach - using a separate shared workflow for the build and then referencing it for packaging and validation.  I wonder if we should make a similar change to the other packages?

File: .github/workflows/bcny-build-nuget-package.yml
Comment (fiedukow) [reply .github/workflows/bcny-build-nuget-package.yml:9]: Can you pass it as a matrix to the callable workflow/action? I'm not aware of this functionality and passing it as a JSON just to parse it on the other end seems like an overkill.

File: .github/workflows/bcny-build-nuget-package.yml
Comment (fiedukow) [reply .github/workflows/bcny-build-nuget-package.yml:43]: I believe it's static. Not sure about non-link linkers as I don't know what they are :)

File: .github/workflows/bcny-build-nuget-package.yml
Comment (fiedukow) [reply .github/workflows/bcny-build-nuget-package.yml]: Sure

File: .github/workflows/bcny-build-nuget-package.yml
Comment (fiedukow) [reply .github/workflows/bcny-build-nuget-package.yml]: Sure

File: .github/workflows/bcny-build-nuget-package.yml
Comment (fiedukow) [reply .github/workflows/bcny-build-nuget-package.yml]: Sure, if we drop all other versions of this lib.

**Review by mikolaj-boc — APPROVED** [review]
Approving to unblock further work

**Comment by fiedukow** [comment]
@compnerd - I'm merging it as is, to unblock further integration. Happy to adjust it more if you have more ideas on how to improve it.

File: .github/workflows/bcny-build-nuget-package.yml
Comment (compnerd) [reply .github/workflows/bcny-build-nuget-package.yml:43]: The only ones that matter are non-link linkers - we use lld for linking Swift code.

Comment (fiedukow) [reply .github/workflows/bcny-build-nuget-package.yml:43]: I really would love to know what `non-link linkers` are, I tried googling that, but there is no combination of this words on the internet according to Google, so I'm kinda confused.

image

Comment (compnerd) [reply .github/workflows/bcny-build-nuget-package.yml:43]: I was referring to lld (as I mentioned in https://github.com/thebrowsercompany/libsodium/pull/2#discussion_r1462019546), but BFD is another. I was being lazy and didn't quote it properly, I meant non-`link` linkers, aka, linkers other than `link.exe` which is Microsoft's linker.

Comment (fiedukow) [reply .github/workflows/bcny-build-nuget-package.yml:43]: I see, I guess it doesn't matter for the `ltcg` configuration as we are not shipping it, but I was successful in linking that `static` as well as `dynamic` with `lld`.

Comment (compnerd) [reply .github/workflows/bcny-build-nuget-package.yml:43]: Right, the issue with the LTCG is that it depends on the compiler and linker working in consort, so built with LTCG may not be possible to use with clang/lld. While using LTCG would be amazing, I don't think that we are in a position to do that currently. If were packaging it, we should ensure that it was usable in the build as it would trip up people otherwise.

Comment (fiedukow) [reply .github/workflows/bcny-build-nuget-package.yml:43]: Makes sense! Thank you for the explanation!
As you requested I limited the nuget package to `static` configuration only, so this is non-issue for now.

**Pull request merged**
